crates/accelerate/src/remove_diagonal_gates_before_measure.rs

```diff
@@ -0,0 +1,103 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+/// Remove diagonal gates (including diagonal 2Q gates) before a measurement.
+use hashbrown::HashSet;
+use pyo3::prelude::*;
+
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::operations::Operation;
+use qiskit_circuit::operations::StandardGate;
+
+/// Run the RemoveDiagonalGatesBeforeMeasure pass on `dag`.
+/// Args:
+///     dag (DAGCircuit): the DAG to be optimized.
+/// Returns:
+///     DAGCircuit: the optimized DAG.
+#[pyfunction]
+#[pyo3(name = "remove_diagonal_gates_before_measure")]
+fn run_remove_diagonal_before_measure(dag: &mut DAGCircuit) -> PyResult<()> {
+    static DIAGONAL_1Q_GATES: [StandardGate; 8] = [
```

Comment:
Would it be faster if these were a `HashSet` and the lookup operation is `O(1)`? We'd have an overhead to construct the HashSet, but we do check `contains` many times 🤔

Reply:
with `HashSet` the time is slightly higher than with static arrays.

| n_qubits | depth | time is sec with static arrays | time in sec with `HashSet` | 
|----------|--------|---------------------------|-------------------------|
| 5 | 1024 | 0.00 | 0.00 |  
| 14 | 1024 | 0.02 | 0.03 | 
| 20 | 1024 | 0.04 | 0.05 |
| 50 | 1024 |  0.31 | 0.32 | 
| 100 | 1024 | 1.52 | 1.54 |

Reply:
nice, thanks for checking! 🙂